src/stream/gzip.rs

```diff
@@ -167,3 +167,150 @@ fn get_header(level: Compression) -> Bytes {
 
     Bytes::from(header)
 }
+
+#[derive(Debug)]
+enum DeState {
+    Reading,
+    Writing,
+    CheckingFooter,
+    Invalid,
+}
+
+#[unsafe_project(Unpin)]
+pub struct DecompressedGzipStream<S: Stream<Item = Result<Bytes>>> {
+    #[pin]
+    inner: S,
+    state: DeState,
+    header_read: bool,
+    input: BytesMut,
+    output: BytesMut,
+    crc: Crc,
+    decompress: Decompress,
+}
+
+impl<S: Stream<Item = Result<Bytes>>> Stream for DecompressedGzipStream<S> {
+    type Item = Result<Bytes>;
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Result<Bytes>>> {
+        let mut this = self.project();
+
+        fn decompress(
+            decompress: &mut Decompress,
+            input: &mut BytesMut,
+            output: &mut BytesMut,
+            crc: &mut Crc,
+            flush: FlushDecompress,
+        ) -> Result<(Status, Bytes)> {
+            const OUTPUT_BUFFER_SIZE: usize = 8_000;
+
+            if output.len() < OUTPUT_BUFFER_SIZE {
+                output.resize(OUTPUT_BUFFER_SIZE, 0);
+            }
+
+            let (prior_in, prior_out) = (decompress.total_in(), decompress.total_out());
+            let status = decompress.decompress(input, output, flush)?;
+            let input_len = decompress.total_in() - prior_in;
+            let output_len = decompress.total_out() - prior_out;
+
+            crc.update(&output[0..output_len as usize]);
+            input.advance(input_len as usize);
+            Ok((status, output.split_to(output_len as usize).freeze()))
+        }
+
+        #[allow(clippy::never_loop)] // https://github.com/rust-lang/rust-clippy/issues/4058
+        loop {
+            break match dbg!(mem::replace(this.state, DeState::Invalid)) {
+                DeState::Reading => {
+                    *this.state = match ready!(this.inner.as_mut().poll_next(cx)) {
+                        Some(chunk) => {
+                            this.input.extend_from_slice(&chunk?);
+                            if !*this.header_read && this.input.len() >= 10 {
```

Comment:
If you split this off to an initial `DeState::ReadingHeader` state then you can restrict the `extend_from_slice` to happening there and avoid the extra copying while reading the rest of the data.

Reply:
Oh, true, I'll try that.